Core/Scripts: Migrate Shadowmoon Valley NPC's to SAI

mysterious-static opened the pull request:
**Changes proposed**:
- Migrates Karynaku (Netherwing quest giver elite dragon thing) to SAI.
- Migrates Drake Dealer Hurlunk (Netherwing drake vendor) to DB (condition vs in script).
- Migrates Flanis Swiftwing and Kagrosh (dead dudes that you get a quest item from) to SAI. This one fixes some non-blizzlike behaviour and migrates gossip texts to DB as well.

I don't think there's anything else easily migrateable here, everything else has more complex scripting unless I missed something obvious.

**Target branch(es)**: 3.3.5

updates velinath/TrinityCore#2, velinath/TrinityCore#5

**Tests performed**:
- I tested Flanis and Kagrosh and they both work beautifully. Karynaku also seems to do everything she's supposed to as far as actions on quest acceptance. Hurlunk also also seems to work as intended. Yay!
- As with the last PR, I don't exactly have a live account to test against, so if anyone wants to do that, feel free, although I'm not sure there's much point.

Commit:
Migrate Flanis Swiftwing and Kagrosh to SAI. More to come.

Commit:
Remove Hurlunk's script. Worked for me(tm)

Commit:
formatting typos bleah

Commit:
Karynaku to SAI. I'm confident enough.

Commit:
NEWLINE

Commit:
I love Kiera2 but I also find my comments more informative.

velinath commented:
The following would be an alternative SQL file if you would prefer it be more compact. The current one separates out all of the NPC's into their own sections for readability.

```sql
-- Add sniffed gossip text to menu
DELETE FROM `gossip_menu_option` WHERE `menu_id` IN (8356, 8371); -- Flanis Swiftwing, Kagrosh

INSERT INTO `gossip_menu_option` (`menu_id`,`id`,`option_icon`,`option_text`,`OptionBroadcastTextID`,`option_id`,`npc_option_npcflag`,`action_menu_id`,`action_poi_id`,`box_coded`,`box_money`,`box_text`,`BoxBroadcastTextID`) VALUES
(8356, 0, 0, 'Reclaim Flanis\'s pack.', 19326,1,1,0,0,0,0,'',0),
(8371, 0, 0, 'Search the corpse for Kagrosh\'s pack.', 19400,1,1,0,0,0,0,'',0);

-- Migrate NPCs to use SmartAI / remove from C++ script
UPDATE `creature_template` SET `AIName` = 'SmartAI', `ScriptName` = '' WHERE `entry` IN (21727, 21725, 22112);
UPDATE `creature_template` SET `ScriptName` = '' WHERE `entry` = 23489;

-- Create SmartAI
DELETE FROM `smart_scripts` WHERE (source_type = 0 AND entryorguid IN (21727, 21725, 22112);
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(21727, 0, 0, 1, 62, 0, 100, 0, 8356, 0, 0, 0, 56, 30658, 1, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Gossip Option 0 Selected - Add Item \'Flanis\'s Pack\'  1 Time'),
(21727, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Linked Action - Close Gossip'), -- fixes current non-Blizzlike behaviour
(21725, 0, 0, 1, 62, 0, 100, 0, 8371, 0, 0, 0, 56, 30659, 1, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Gossip Option 0 Selected - Add Item \'Kagrosh\'s Pack\'  1 Time'),
(21725, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Linked Action - Close Gossip'), -- fixes current non-Blizzlike behaviour
(22112, 0, 0, 0, 19, 0, 100, 0, 10866, 0, 0, 0, 12, 11980, 1, 300000, 0, 0, 0, 8, 0, 0, 0, -4204.94, 316.397, 122.508, 1.309, 'Karynaku - On Quest Accepted - Summon Creature \'Zuluhed the Whacked\''),
(22112, 0, 1, 0, 19, 0, 100, 0, 10870, 0, 0, 0, 52, 649, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Karynaku - On Quest Accepted - Activate Taxi Path 649');

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` IN (8356, 8371, 8754);
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(15,8356,0,0,0,2,0,30658,1,0,1,0,'','Flanis Swiftwing - Show gossip option 0 if player does not have item 30658 in inventory'),
(15,8356,0,0,0,9,0,10583,0,0,0,0,'','Flanis Swiftwing - Show gossip option 0 if player has accepted quest 10583');
(15,8371,0,0,0,2,0,30659,1,0,1,0,'','Kagrosh - Show gossip option 0 if player does not have item 30659 in inventory'),
(15,8371,0,0,0,9,0,10601,0,0,0,0,'','Kagrosh - Show gossip option 0 if player has accepted quest 10601');
(15,8754,0,0,0,5,0,1015,128,0,0,0,'','Drake Dealer Hurlunk - Show gossip option 0 if player is exalted with faction 1015');
```

Aokromes commented:
I prefer the 2nd :)

Commit:
compactify sql

velinath commented:
I'll compact these moving forward, in that case. :+1:

tkrokli commented:
An updated Keira2 based on TDB 335.61 + updates (and also TDB 6.04 + updates)has been set up on https://gnaxi.net/wow/keira2/ (online 24/7) by https://github.com/GnaXi :+1:

Aokromes commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -0,0 +1,27 @@
+(15,8356,0,0,0,9,0,10583,0,0,0,0,'','Flanis Swiftwing - Show gossip option 0 if player has accepted quest 10583'),
+(15,8371,0,0,0,2,0,30659,1,0,1,0,'','Kagrosh - Show gossip option 0 if player does not have item 30659 in inventory'),
+(15,8371,0,0,0,9,0,10601,0,0,0,0,'','Kagrosh - Show gossip option 0 if player has accepted quest 10601'),
+(15,8754,0,0,0,5,0,1015,128,0,0,0,'','Drake Dealer Hurlunk - Show gossip option 0 if player is exalted with faction 1015');
```
newline

velinath replied:
I checked the file in github and it looks like it has a newline already

Aokromes commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -0,0 +1,27 @@
+UPDATE `creature_template` SET `ScriptName` = '' WHERE `entry` = 23489;
+
+-- Create SmartAI
+DELETE FROM `smart_scripts` WHERE (source_type = 0 AND entryorguid IN (21727, 21725, 22112);
```
missing )

Aokromes commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -0,0 +1,27 @@
+(21727, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Linked Action - Close Gossip'),
+(21725, 0, 0, 1, 62, 0, 100, 0, 8371, 0, 0, 0, 56, 30659, 1, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Gossip Option 0 Selected - Add Item \'Kagrosh\'s Pack\'  1 Time'),
+(21725, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Linked Action - Close Gossip'),
+(22112, 0, 0, 0, 19, 0, 100, 0, 10866, 0, 0, 0, 12, 11980, 1, 300000, 0, 0, 0, 8, 0, 0, 0, -4204.94, 316.397, 122.508, 1.309, 'Karynaku - On Quest Accepted - Summon Creature \'Zuluhed the Whacked\''),
```
SmartAIMgr: Entry 21725 SourceType 0 Event 0 Action 56 Create Item: There is a create item spell for item 30659 (SpellId: 37245 effect: 0)
SmartAIMgr: Entry 21727 SourceType 0 Event 0 Action 56 Create Item: There is a create item spell for item 30658 (SpellId: 37244 effect: 0)

velinath replied:
@Aokromes I just used the same logic as the script, should I use the spell with target invoker then?

Aokromes replied:
Don't look at me xd i only known i get those startup errors after i applied the sql.

velinath replied:
I'll test with the spellids later this week and see if the quest still works, can't imagine why it wouldn't though :)

Commit:
syntax fail

Commit:
SAI changes per comment

This still needs testing, but since there's a spell ingame, went ahead and changed this now. I can always revert later if needed.

Aokromes commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -11,9 +11,9 @@ UPDATE `creature_template` SET `ScriptName` = '' WHERE `entry` = 23489;
+(21727, 0, 0, 1, 62, 0, 100, 0, 8356, 0, 0, 0, 11, 37244, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Gossip Option 0 Selected - Add Item \'Flanis\'s Pack\'  1 Time'),
 (21727, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Linked Action - Close Gossip'),
-(21725, 0, 0, 1, 62, 0, 100, 0, 8371, 0, 0, 0, 56, 30659, 1, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Gossip Option 0 Selected - Add Item \'Kagrosh\'s Pack\'  1 Time'),
+(21725, 0, 0, 0, 62, 0, 100, 0, 8371, 0, 0, 0, 11, 37245, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Gossip Option 0 Selected - Cast Spell');
```
; instead ,

velinath replied:
too early :sob:

Commit:
astonishing

thanks for the catch @Aokromes. Early morning edits...

Aokromes commented:
SmartAIMgr::LoadSmartAIFromDB: Entry 21725 SourceType 0, Event 1, Link Source Event not found or invalid. Event will never trigger.

Aokromes commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -11,9 +11,9 @@ UPDATE `creature_template` SET `ScriptName` = '' WHERE `entry` = 23489;
 -- Create SmartAI
 DELETE FROM `smart_scripts` WHERE (source_type = 0 AND entryorguid IN (21727, 21725, 22112));
 INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`tar ...
```
SmartAIMgr::LoadSmartAIFromDB: Entry 21725 SourceType 0, Event 1, Link Source Event not found or invalid. Event will never trigger.

velinath replied:
this is just an embarrassing disaster

Commit:
relink Kagrosh SAI events

velinath commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -0,0 +1,27 @@
+INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`tar ...
+(21727, 0, 0, 1, 62, 0, 100, 0, 8356, 0, 0, 0, 11, 37244, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Gossip Option 0 Selected - Cast Spell'),
+(21727, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Linked Action - Close Gossip'),
+(21725, 0, 0, 1, 62, 0, 100, 0, 8371, 0, 0, 0, 11, 37245, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Kagrosh - On Gossip Option 0 Selected - Cast Spell'),
```
Spellcast does not add item to player inventory

Killyana replied:
Are you testing it on 3.3.5 or 6.x?

velinath replied:
Oh awesome, thanks for the help here. I learned something today!

velinath commented on sql/updates/world/3.3.5/2016_06_14_99_world.sql:
```diff
@@ -0,0 +1,27 @@
+DELETE FROM `smart_scripts` WHERE (source_type = 0 AND entryorguid IN (21727, 21725, 22112));
+INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`tar ...
+(21727, 0, 0, 1, 62, 0, 100, 0, 8356, 0, 0, 0, 11, 37244, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Gossip Option 0 Selected - Cast Spell'),
+(21727, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Flanis Swiftwing - On Linked Action - Close Gossip'),
```
Spellcast does not add item to player inventory

Killyana replied:
Same here, use the cast flag=2 as the npc is dead.

Commit:
update for cast when dead

thanks @Killyana!

Aokromes commented:
Soo now works? :)

velinath commented:
yep, tested, working, good to go from my perspective
